File: OJ/leet1114/leet1114_Condition.py
# leet1114

from typing import *


import threading
from multiprocessing import pool 

printFirst = lambda: print('First')
printSecond = lambda: print('Second')
printThird = lambda: print('Third')

class Foo:
    def __init__(self):
        self.cv = threading.Condition()
        self.t = 0

    def first(self, printFirst: 'Callable[[], None]') -> None:
        self.res(0, printFirst)

    def second(self, printSecond: 'Callable[[], None]') -> None:
        self.res(1, printSecond)

    def third(self, printThird: 'Callable[[], None]') -> None:
        self.res(2, printThird)
        
    def res(self, val: int, func: 'Callable[[], None]') -> None:
        with self.cv:
            self.cv.wait_for(lambda: val == self.t) #参数是函数对象，返回值是bool类型
            func()
            self.t += 1
            self.cv.notify_all()
 


foo = Foo()
tfirst = threading.Thread(target=foo.first, args=(printFirst,))
tsecond = threading.Thread(target=foo.second, args=(printSecond, ))
tthird = threading.Thread(target=foo.third, args=(printThird,))
tthird2 = threading.Thread(target=foo.third, args=(printThird,))
tthird.start()
tsecond.start()
tfirst.start()

File: OJ/leet1114/leet1114_Condition.cpp
/*
leet1114


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
#include <thread>
#include <mutex>
#include <condition_variable>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

void printFirst() { cout << "First\n";  }
void printSecond() {  cout << "Second\n"; }
void printThird() {  cout << "Third\n"; }


class Foo {
public:
    condition_variable cv;
    mutex mx;
    int k = 0;
    
    Foo() { }

    void first(function<void()> printFirst) {
        
        printFirst();

        k = 1;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {

        unique_lock Lock(mx);
        cv.wait(Lock, [&](){ return k == 1; } );

        printSecond();

        k = 2;
        cv.notify_all();
    }

    void third(function<void()> printThird) {

        unique_lock Lock(mx);
        cv.wait(Lock, [&](){ return k == 2; } );

        printThird();
        
    }
};


int main(int argc, char const *argv[])
{
    Foo foo;
    thread t1(&Foo::first, &foo, printFirst);
    thread t2(&Foo::second, &foo, printSecond);
    thread t3(&Foo::third, &foo, printThird);
    t3.join();
    t2.join();
    t1.join();
    return 0;
}
